test.js

```diff
@@ -0,0 +1,594 @@
+  t.plan(9)
+
+  const server = createServer()
+  server.on('close', () => t.pass('server closed'))
+
+  server.on('connection', function (socket) {
+    socket.on('close', () => t.pass('server socket closed'))
+    socket.on('error', (err) => t.fail('server socket error: ' + err.message + ' (' + err.code + ')'))
+  })
+
+  server.on('request', function (req, res) {
+    res.writeHead(404, { 'x-custom': 1234 }) // + should set content-length to zero? otherwise the client receives "transfer-enconding: chunked"
+    res.end()
+
+    req.on('close', () => t.pass('server request closed'))
+    res.on('close', () => t.pass('server response closed'))
+  })
+
+  server.listen(0)
+  await waitForServer(server)
+
+  const req = http.request({
+    method: 'GET',
+    host: server.address().address,
+    port: server.address().port,
+    path: '/'
+  }, function (res) {
+    t.is(res.statusCode, 404)
+
+    const customHeader = res.rawHeaders.indexOf('X-Custom')
+    t.is(res.rawHeaders[customHeader], 'X-Custom')
+    t.is(res.rawHeaders[customHeader + 1], '1234')
+
+    res.on('data', () => t.fail('client should not receive data'))
+    res.on('end', () => t.pass('client response ended'))
+  })
+
+  req.on('close', () => {
+    t.pass('client request closed')
+    server.close()
+  })
+
+  req.on('error', (err) => t.fail('client req error: ' + err.message + ' (' + err.code + ')'))
+
+  req.end()
+})
+
+test('chunked', async function (t) {
+  t.plan(9)
+
+  const server = createServer()
+  server.on('close', () => t.pass('server closed'))
+
+  server.on('connection', function (socket) {
+    socket.on('close', () => t.pass('server socket closed'))
+    socket.on('error', (err) => t.fail('server socket error: ' + err.message + ' (' + err.code + ')'))
+  })
+
+  server.on('request', function (req, res) {
+    res.write('part 1 + ')
+    setImmediate(() => {
+      res.end('part 2')
+    })
+
+    t.is(res.chunked, true, 'chunked by default')
+    res.on('close', () => {
+      t.is(res.chunked, true, 'still chunked')
+    })
+
+    req.on('close', () => t.pass('server request closed'))
+    res.on('close', () => t.pass('server response closed'))
+  })
+
+  server.listen(0)
+  await waitForServer(server)
+
+  const req = http.request({
+    method: 'GET',
+    host: server.address().address,
+    port: server.address().port,
+    path: '/'
+  }, function (res) {
+    t.is(res.statusCode, 200)
+
+    const chunks = []
+    res.on('data', (chunk) => chunks.push(chunk))
+    res.on('end', () => {
+      const body = b4a.concat(chunks)
+      t.alike(body, b4a.from('part 1 + part 2'), 'client response ended')
+    })
+  })
+
+  req.on('close', () => {
+    t.pass('client request closed')
+    server.close()
+  })
+
+  req.on('error', (err) => t.fail('client req error: ' + err.message + ' (' + err.code + ')'))
+
+  req.end()
+})
+
+// + solo this test and it would randomly fail
+test('server does a big write', async function (t) {
```

Comment:
Solo this test and it would randomly fail

Reply:
https://github.com/holepunchto/tiny-http-native/pull/3/files#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R219